docs/specs/component.md

```diff
@@ -0,0 +1,65 @@
+# Component Specification
+
+This document specifies Vector Component behavior (source, transforms, and
+sinks) for the development of Vector.
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”,
+“SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be
+interpreted as described in [RFC 2119].
+
+<!-- MarkdownTOC autolink="true" style="ordered" indent="   " -->
+
+1. [Introduction](#introduction)
+1. [How to read this document](#how-to-read-this-document)
+1. [Instrumentation](#instrumentation)
+   1. [Events](#events)
+      1. [EventRecevied](#eventrecevied)
+
+<!-- /MarkdownTOC -->
+
+## Introduction
+
+Vector is a highly flexible observability data pipeline due to its directed
+acyclic graph processing model. Each node in the graph is a Vector Component,
+and in order to meet our [high user experience expectations] each Component must
+adhere to a common set of behaviorial rules. This document aims to clearly
+outline these rules to guide new component development and ongoing maintenance.
+
+## How to read this document
+
+This document is written from the broad perspective of a Vector component.
+Unless otherwise stated, a section applies to all component types, although,
+most sections will be broken along component lines for easy adherence.
+
+## Instrumentation
+
+### Events
+
+Vector implements an [event driven pattern ([RFC 2064]) for internal
+instrumentation. This section lists all required and optional events that a
+component MUST emit.
+
```

Comment:
Also worth noting that all errors should emit an `Event`, and it should increment an error counter (ie `processing_errors_total`) with tags.

Reply:
Agreed, we'll want to discuss errors in this doc too.

Reply:
Yep, I'm on it.